R/env-special.R

```diff
@@ -197,44 +197,51 @@ current_env <- function() {
 #' These functions are experimental and may not belong to the rlang
 #' package. Expect API changes.
 #'
-#' @param pkg The name of a package. If `NULL`, the surrounding
-#'   namespace is returned, or an error is issued if not called within
-#'   a namespace. If a function, the enclosure of that function is
-#'   checked.
+#' @param pkg The name of a package or an environment. If a function,
+#'   the environment of that function is searched for a namespace with
+#'   [base::topenv()].
 #'
 #' @seealso [pkg_env()]
 #' @keywords internal
 #' @export
 ns_env <- function(pkg = NULL) {
-  if (is_null(pkg)) {
-    bottom <- topenv(caller_env())
-    if (!isNamespace(bottom)) abort("not in a namespace")
-    bottom
-  } else if (is_function(pkg)) {
-    env <- env_parent(pkg)
-    if (isNamespace(env)) {
-      env
-    } else {
-      NULL
-    }
-  } else {
-    asNamespace(pkg)
+  env <- switch(typeof(pkg),
+    builtin = ,
+    special =
+      ns_env("base"),
+    closure =
+      topenv(fn_env(pkg)),
+    environment =
+      pkg,
+    character =
+      if (is_string(pkg)) {
+        asNamespace(pkg)
+      }
+  )
+
+  if (!is_namespace(env)) {
+    abort("`pkg` must be a package name or a function inheriting from a namespace.")
   }
+
+  env
 }
 #' @rdname ns_env
 #' @export
 ns_imports_env <- function(pkg = NULL) {
   env_parent(ns_env(pkg))
 }
 #' @rdname ns_env
+#' @param env A namespace environment.
 #' @export
 ns_env_name <- function(pkg = NULL) {
-  if (is_null(pkg)) {
-    pkg <- with_env(caller_env(), ns_env())
-  } else if (is_function(pkg)) {
-    pkg <- get_env(pkg)
-  }
-  unname(getNamespaceName(pkg))
+  env <- switch(typeof(pkg),
```

Comment:
Why not use `ns_env()` here?

Reply:
Because `ns_env()` has a different error message.